File: 2021/day3/scala/solution.scala
import scala.io.Source

@main
def main() = {
  println("Part 1: ")
  print("Example: ")
  part1("../example.txt")
  print("Input: ")
  part1("../input.txt")
  println()

  println("Part 2: ")
  print("Example: ")
  part2("../example.txt")
  print("Input: ")
  part2("../input.txt")
  println()
}

def parseInput(path: String) = {
  Source.fromFile(path)
  .getLines
  .toList
}

def binaryStringToInt(binary : String) =
  val (_, result) = binary.foldRight((0,0))((char, v) => {
    val (index, value) = v
    if (char == '1')
      (index + 1, value + math.pow(2, index).toInt)
    else
      (index + 1, value)
  })
  result

def part1(path: String) = {
  val input = parseInput(path)

  val lineCount = input.length
  val lineLength = input(0).length
  val count = Array.fill[Int](lineLength){0}

  for(line <- input; i <- (0 to lineLength - 1))
    if line(i) == '1' then
      count(i) = count(i) + 1

  var gamma = ""
  var epsilon = ""
  for(v <- count)
    if v > (lineCount / 2) then
      gamma = gamma + "1"
      epsilon = epsilon + "0"
    else
      gamma = gamma + "0"
      epsilon = epsilon + "1"

  val gammaVal = binaryStringToInt(gamma)
  val epsilonVal = binaryStringToInt(epsilon)
  val result = gammaVal * epsilonVal
  println(s"Result: $result")
}

def determineRatingByPrio(input: List[String], priority : Char, offset : Int = 0) : String = {
  if (input.length == 1)
    return input(0)

  if (input.length == 2)
    if input(0)(offset) == priority then
      return input(0)
    else
      return input(1)

  val (countZero, countOne) = input.foldLeft((0,0))((state, line) => {
    line(offset) match
      case '0' => (state._1 + 1, state._2)
      case '1' => (state._1, state._2 + 1)
  })

  val searchFor = priority match
    case '0' => if countZero <= countOne then '0' else '1'
    case '1' => if countOne >= countZero then '1' else '0'

  determineRatingByPrio(
    input.filter(it => it(offset) == searchFor),
    priority,
    offset + 1)    
}

def part2(path: String) = {
  val input = parseInput(path)

  val oxy = determineRatingByPrio(input, '1')
  val co2 = determineRatingByPrio(input, '0')
  println()
  println(s"oxy = $oxy = ${binaryStringToInt(oxy)}")
  println(s"co2 = $co2 = ${binaryStringToInt(co2)}")

  println(binaryStringToInt(oxy) * binaryStringToInt(co2))
}

File: 2021/day3/ruby/solution.rb
input = File.readlines(__dir__ + '/../input.txt').map(&:strip)
example = File.readlines(__dir__ + '/../example.txt').map(&:strip)

def bit_counting(input)
  input
    .map { _1.split('') }
    .each_with_object(input.first.size.times.map { { zero: 0, one: 0 } }) do |binary, counter|
    binary.each_with_index do |value, index|
      case value
      in '1' then counter[index][:one] += 1
      in '0' then counter[index][:zero] += 1
      end
    end
  end
end

def gamma_rate(input)
  bit_counting(input).map { _1[:zero] > _1[:one] ? '0' : '1' }.join('').to_i(2)
end

def epsilon_rate(input)
  bit_counting(input).map { _1[:zero] > _1[:one] ? '1' : '0' }.join('').to_i(2)
end

def oxygen_generator_rating(input, look_at_index = 0)
  counted_bits = bit_counting(input)
  new_input = input.filter do |line|
    if counted_bits[look_at_index][:one] >= counted_bits[look_at_index][:zero]
      line[look_at_index] == '1'
    else
      line[look_at_index] == '0'
    end
  end
  return new_input.first.to_i(2) if new_input.size == 1

  oxygen_generator_rating(new_input, look_at_index + 1)
end

def co2_scrubber_rating(input, look_at_index = 0)
  counted_bits = bit_counting(input)
  new_input = input.filter do |line|
    if counted_bits[look_at_index][:zero] <= counted_bits[look_at_index][:one]
      line[look_at_index] == '0'
    else
      line[look_at_index] == '1'
    end
  end
  return new_input.first.to_i(2) if new_input.size == 1

  co2_scrubber_rating(new_input, look_at_index + 1)
end

def part1(input)
  gamma_rate(input) * epsilon_rate(input)
end

def part2(input)
  oxygen_generator_rating(input) * co2_scrubber_rating(input)
end

puts "Part 1 (Example): #{part1(example)}"
puts "Part 1: #{part1(input)}"
puts "Part 2 (Example): #{part2(example)}"
puts "Part 2: #{part2(input)}"
